File: september/week1/all_elements_in_two_binary_search_trees.cpp
#include <bits/stdc++.h>

using namespace std;
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> getAllElements(TreeNode* root1, TreeNode* root2) {
        vector<int> ret;
        stack<TreeNode *> s;
        s.push(root1);
        while(!s.empty()){
            TreeNode *cur = s.top();
            s.pop();
            if(cur==nullptr)  continue;
            ret.push_back(cur->val);
            s.push(cur->left);
            s.push(cur->right);
        }
        s.push(root2);
        while(!s.empty()){
            TreeNode *cur = s.top();
            s.pop();
            if(cur==nullptr)  continue;
            ret.push_back(cur->val);
            s.push(cur->left);
            s.push(cur->right);
        }
        sort(ret.begin(),ret.end());
        return ret;
    }
};

File: september/week1/all_elements_in_two_binary_search_trees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def inorder(self, root: TreeNode) -> List:
        return self.inorder(root.left) + [root.val] + self.inorder(root.right) if root else []
    def getAllElements(self, root1: TreeNode, root2: TreeNode) -> List[int]:
        arr1 = self.inorder(root1)
        arr2 = self.inorder(root2)
        n = len(arr1)
        m = len(arr2)
        i = j = 0
        ret = []
        while i < n and j < m:
            if arr1[i]<arr2[j]:
                ret.append(arr1[i])
                i += 1
            else:
                ret.append(arr2[j])
                j += 1
        while i < n:
            ret.append(arr1[i])
            i += 1
        while j < m:
            ret.append(arr2[j])
            j += 1
        return ret
        
